crypto/pkcs7/pkcs7_asn1.c

```diff
@@ -0,0 +1,186 @@
+#include <openssl/pkcs7.h>
+
+#include <openssl/asn1t.h>
+#include <openssl/pem.h>
+
+#include "internal.h"
+#include "../internal.h"
+#include "../bytestring/internal.h"
+
+ASN1_ADB_TEMPLATE(p7default) = ASN1_EXP_OPT(PKCS7, d.other, ASN1_ANY, 0);
+
+ASN1_ADB(PKCS7) = {
+        ADB_ENTRY(NID_pkcs7_data, ASN1_EXP_OPT(PKCS7, d.data, ASN1_OCTET_STRING, 0)),
+        ADB_ENTRY(NID_pkcs7_signed, ASN1_EXP_OPT(PKCS7, d.sign, PKCS7_SIGNED, 0)),
+        ADB_ENTRY(NID_pkcs7_enveloped, ASN1_EXP_OPT(PKCS7, d.enveloped, PKCS7_ENVELOPE, 0)),
+        ADB_ENTRY(NID_pkcs7_signedAndEnveloped, ASN1_EXP_OPT(PKCS7, d.signed_and_enveloped, PKCS7_SIGN_ENVELOPE, 0)),
+        ADB_ENTRY(NID_pkcs7_digest, ASN1_EXP_OPT(PKCS7, d.digest, PKCS7_DIGEST, 0)),
+        ADB_ENTRY(NID_pkcs7_encrypted, ASN1_EXP_OPT(PKCS7, d.encrypted, PKCS7_ENCRYPT, 0))
+} ASN1_ADB_END(PKCS7, 0, type, 0, &p7default_tt, NULL);
+
+ASN1_SEQUENCE(PKCS7) = {
+    ASN1_SIMPLE(PKCS7, type, ASN1_OBJECT),
+    ASN1_ADB_OBJECT(PKCS7)
+} ASN1_SEQUENCE_END(PKCS7)
+
+IMPLEMENT_ASN1_ALLOC_FUNCTIONS(PKCS7)
+
+PKCS7 *d2i_PKCS7(PKCS7 **a, const unsigned char **in, long len) {
```

Comment:
Just a minor nit, would it be worth adding a TODO comment here stating that we want to translate all of this CBB/CBS in some distant future

Reply:
I don't think so. Personally, I prefer to keep committed TODO's much more specific and narrower in scope.